### PR TITLE
Prefer Option over Some

### PR DESCRIPTION
**Pull request by kionka**

### CODE REVIEW

**Comment by joewalp** [comment]
I'd put it at the top of the Option section.

**Comment by ayakushev99** [comment]
Wrapping null is only one of possible uses of `Option`. In functions like `Iterator.find(p: T => Boolean)` the `Option` return type represents whether or not the corresponding value was matched (even if it is `null`). While I agree that the vast majority of the time `Some(x)` should not be used, we should clarify that there are exceptions.

**Comment by milesoc** [comment]
Even in that case you wouldn't wrap something in Some(). Maybe we should clarify that this doesn't apply to ```Some.unapply``` which is reasonable in case matching.

**Comment by ayakushev99** [comment]
How I would write this:
When creating an `Option` you should use `Option(x)` constructor, not `Some(x)`. The only exception to this is if you are sure that returning a value of `Some(null)` is acceptable for your use case. An example of this is `Iterator.find(p: T => Boolean): Option[T]`, where the `Option` type represents whether the matching value exists, and doesn't care whether it's `null` or not. Note, that in these exceptional cases, you should end up with a code block that returns either `Some(x)` or `None`, so the resulting inferred type of the code block will be `Option[T]`

For wrapping constants `.some` may be used.

**Comment by ayakushev99** [comment]
@milesoc yes, you would. The `Iterator.find` function should not care if some values are null. Again, it uses `Option` for a different purpose. If your collection may contain nulls, and you want to ensure that the result doesn't come back as `Some(null)` you should pre-wrap, or pre-filter the contents of the collection.

**Comment by milesoc** [comment]
@ayakushev99 then you would wrap the collection with ```Option()``` and still wouldn't use ```Some```

**Comment by milesoc** [comment]
I see what you're saying, for when writing ```find```. I honestly don't think that that should be allowed - to me if you're going to return ```Option``` the end-user should be able to count on it not being ```Option(null)```. For the sake of our codebase we should also be able to assume that collections don't contain ```null```, and if they might, you probably should have filtered through it beforehand (for example coming from java)

**Comment by 2rs2ts** [comment]
I'm inclined to agree with @ayakushev99 but I think we can also write down "always use `Option`" but in practice understand that in some rare circumstances we should use `Some` and break our rules.

File: README.md [hunk 1]
```diff
@@ -402,6 +402,27 @@ Option(123) match {
...
+instead, which throws on `null`.
+
+```scala
+val doThisForConstants = "hello".some
+val notThisForConstants = "goodbye".opt
+
+val doThisForEverythingElse = foo().opt
+val notThisForEverythingElse = bar().some
+```
+
```
Comment (arschles) [inline README.md]: this is a lot to read for a simple rule. I'd like to simplify:

use `Option(...)` or [`.opt`](https://github.com/paypal/cascade/blob/develop/common/src/main/scala/com/paypal/cascade/common/option/option.scala#L61) to create `Option`s. Only use `Some(...)` or [`.some`](https://github.com/paypal/cascade/blob/develop/common/src/main/scala/com/paypal/cascade/common/option/option.scala#L55) for constant, non-null values. Avoid methods and styles that will return `Some(null)`, such as `Iterator.find(p: T => Boolean)`

**Comment by arschles** [comment]
LGTM

**Pull request closed without merging**